src/connmgr/connection.rs

```diff
@@ -5185,30 +5183,29 @@ where
 
                         match ret {
                             RecvStatus::Complete(finished, _) => break finished,
-                            RecvStatus::Read((), written) => {
-                                if written == 0 {
-                                    let mut add_to_buffer = pin!(resp_body.add_to_buffer());
-
-                                    loop {
-                                        // ABR: select contains read
-                                        let result = select_2(
-                                            add_to_buffer.as_mut(),
-                                            pin!(zsess_in.recv_msg()),
-                                        )
-                                        .await;
-
-                                        match result {
-                                            Select2::R1(ret) => {
-                                                ret?;
-                                                break;
-                                            }
-                                            Select2::R2(ret) => {
-                                                let zreq = ret?;
-
-                                                // ABR: handle_other
-                                                server_handle_other(zreq, zsess_in, zsess_out)
-                                                    .await?;
-                                            }
+                            RecvStatus::Read((), size) => {
+                                // buf is non-empty so this can never be zero
+                                assert!(size > 0);
+                            }
+                            RecvStatus::NeedBytes(()) => {
+                                let mut add_to_buffer = pin!(resp_body.add_to_buffer());
+
+                                loop {
+                                    // ABR: select contains read
+                                    let result =
+                                        select_2(add_to_buffer.as_mut(), pin!(zsess_in.recv_msg()))
+                                            .await;
+
+                                    match result {
+                                        Select2::R1(ret) => {
+                                            ret?;
+                                            break;
+                                        }
+                                        Select2::R2(ret) => {
+                                            let zreq = ret?;
+
+                                            // ABR: handle_other
+                                            server_handle_other(zreq, zsess_in, zsess_out).await?;
                                         }
```

Comment:
This is similar to earlier code that reads response body bytes from a server into a buffer that is guaranteed to be non-zero-sized. The updates have identical behavior as before so there should be no performance impact.